enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java

```diff
@@ -30,76 +32,130 @@
+        users = new HashMap<>(numUsers);
         for (int i = 0; i < numUsers; i++) {
-            users.add(in.readString());
+            String userName = in.readString();
+            SecureHash secureHash = in.readOptionalWriteable(SecureHash::readFrom);
+            users.put(userName, secureHash);
         }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        out.writeStringList(users);
+        out.writeVInt(users.size());
+        for (Map.Entry<String, SecureHash> user : users.entrySet()) {
+            out.writeString(user.getKey());
+            out.writeOptionalWriteable(user.getValue());
+        }
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startArray("users");
-        for (String user : users) {
-            builder.value(user);
+        builder.startObject("users");
+        for (Map.Entry<String, SecureHash> entry : users.entrySet()) {
+            builder.startObject(entry.getKey());
+            if (entry.getValue() != null) {
+                entry.getValue().toXContent(builder, params);
+            }
+            builder.endObject();
         }
-        builder.endArray();
+        builder.endObject();
         return builder;
     }
 
+    /**
+     * UsersMetaData v2 has the form of:
+     *
+     * users: {
+     *   "user1": {
+     *     "secure_hash": {
+     *       "iterations": INT,
+     *       "hash": BYTE[],
+     *       "salt": BYTE[]
+     *     }
+     *   },
+     *   "user2": {
+     *     "secure_hash": null
+     *   },
+     *   ...
+     * }
+     *
+     * UsersMetaData v1 has the form of:
+     *
+     * users: [
+     *   "user1",
+     *   "user2",
+     *   ...
+     * ]
+     *
+     */
     public static UsersMetaData fromXContent(XContentParser parser) throws IOException {
-        List<String> users = new ArrayList<>();
-        if (parser.nextToken() == XContentParser.Token.FIELD_NAME && parser.currentName().equals("users")) {
-            if (parser.nextToken() == XContentParser.Token.START_ARRAY) {
-                XContentParser.Token token;
+        Map<String, SecureHash> users = new HashMap<>();
+        XContentParser.Token token = parser.nextToken();
+
+        if (token == XContentParser.Token.FIELD_NAME && parser.currentName().equals("users")) {
+            token = parser.nextToken();
+            if (token == XContentParser.Token.START_ARRAY) {
+                // UsersMetaData v1
                 while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY && token != null) {
-                    users.add(parser.text());
+                    users.put(parser.text(), null); // old users do not have passwords
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                // UsersMetaData v2
+                while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
```

Comment:
I thought we wanted to prevent such conditions in favor of a proper migration *before* reading any user data.

Reply:
@mxm not sure if I get you right here. However I think it's fine if we apply the migration if any cluster state change occurs.

Reply:
It would be nice to have just one way of reading user data and migrating older formats of the user data with the migration service. @chaudum explained it's not really possible to perform the migration before reading any user meta data. So I guess we have to leave it as it is.